core/profiling/stacksampling/src/sampler.cpp

```diff
@@ -0,0 +1,380 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#include "CorProfiler.h"
+#include "sampler.h"
+#include <thread>
+#include <cwchar>
+#include <cstdio>
+#include <cinttypes>
+#include <locale>
+#include <codecvt>
+
+using std::wstring_convert;
+using std::codecvt_utf8;
+using std::string;
+
+ManualEvent Sampler::s_waitEvent;
+Sampler *Sampler::s_instance = nullptr;
+
+HRESULT __stdcall DoStackSnapshotStackSnapShotCallbackWrapper(
+    FunctionID funcId,
+    UINT_PTR ip,
+    COR_PRF_FRAME_INFO frameInfo,
+    ULONG32 contextSize,
+    BYTE context[],
+    void* clientData)
+{
+    return Sampler::Instance()->StackSnapshotCallback(funcId,
+        ip,
+        frameInfo,
+        contextSize,
+        context,
+        clientData);
+}
+
+Sampler::Sampler(ICorProfilerInfo10* pProfInfo, CorProfiler *parent) :
+    m_workerThread(DoSampling, pProfInfo, parent)
+{
+    Sampler::s_instance = this;
+}
+
+// static
+void Sampler::DoSampling(ICorProfilerInfo10 *pProfInfo, CorProfiler *parent)
+{
+    Sampler::Instance()->corProfilerInfo = parent->corProfilerInfo;
+
+    pProfInfo->InitializeCurrentThread();
+
+    while (true)
+    {
+        std::this_thread::sleep_for(std::chrono::milliseconds(100));
+
+        s_waitEvent.Wait();
+
+        if (!parent->IsRuntimeExecutingManagedCode())
```

Comment:
I'd recommend adding some comments around this to clarify:
1. This isn't required for correctness, it was convenient for demonstration purposes
2. The implementation of this call isn't a precise way to determine if managed code is running because the runtime can always invoke pre-compiled code before jitting. However for a sample it was an easy to implement heuristic.

Reply:
Thanks!